sdk/core/azure-core-amqp/src/models/amqp_message.cpp

```diff
@@ -378,6 +378,291 @@ namespace Azure { namespace Core { namespace Amqp { namespace Models {
+    {
+      AmqpMap appProperties;
+      for (auto const& val : message.ApplicationProperties)
+      {
+        if ((val.second.GetType() == AmqpValueType::List)
+            || (val.second.GetType() == AmqpValueType::Map)
+            || (val.second.GetType() == AmqpValueType::Composite)
+            || (val.second.GetType() == AmqpValueType::Described))
+        {
+          throw std::runtime_error(
+              "Message Application Property values must be simple value types");
+        }
+        appProperties.emplace(val);
+      }
+      AmqpValue propertiesValue{amqpvalue_create_application_properties(AmqpValue{appProperties})};
+      auto serializedApplicationProperties = AmqpValue::Serialize(propertiesValue);
+      rv.insert(
+          rv.end(), serializedApplicationProperties.begin(), serializedApplicationProperties.end());
+    }
+
+    switch (message.BodyType)
+    {
+      default:
+      case MessageBodyType::Invalid:
+        throw std::runtime_error("Invalid message body type.");
+
+      case MessageBodyType::Value: {
+        // The message body element is an AMQP Described type, create one and serialize the
+        // described body.
+        AmqpDescribed describedBody(
+            static_cast<std::uint64_t>(AmqpDescriptors::DataAmqpValue), message.m_amqpValueBody);
+        auto serializedBodyValue = AmqpValue::Serialize(describedBody);
+        rv.insert(rv.end(), serializedBodyValue.begin(), serializedBodyValue.end());
+      }
+      break;
+      case MessageBodyType::Data:
+        for (auto const& val : message.m_binaryDataBody)
+        {
+          AmqpDescribed describedBody(static_cast<std::uint64_t>(AmqpDescriptors::DataBinary), val);
+          auto serializedBodyValue = AmqpValue::Serialize(describedBody);
+          rv.insert(rv.end(), serializedBodyValue.begin(), serializedBodyValue.end());
+        }
+        break;
+      case MessageBodyType::Sequence: {
+        for (auto const& val : message.m_amqpSequenceBody)
+        {
+          AmqpDescribed describedBody(
+              static_cast<std::uint64_t>(AmqpDescriptors::DataAmqpSequence), val);
+          auto serializedBodyValue = AmqpValue::Serialize(describedBody);
+          rv.insert(rv.end(), serializedBodyValue.begin(), serializedBodyValue.end());
+        }
+      }
+    }
+    if (!message.Footer.empty())
+    {
+      AmqpValue footer{amqpvalue_create_footer(AmqpValue{message.Footer})};
+      auto serializedFooter = AmqpValue::Serialize(footer);
+      rv.insert(rv.end(), serializedFooter.begin(), serializedFooter.end());
+    }
+
+    return rv;
+  }
+
+  // The message fields, in their expected order.
+  std::vector<Amqp::_detail::AmqpDescriptors> expectedMessageFields{
+      AmqpDescriptors::Header,
+      AmqpDescriptors::DeliveryAnnotations,
+      AmqpDescriptors::MessageAnnotations,
+      AmqpDescriptors::Properties,
+      AmqpDescriptors::ApplicationProperties};
+
+  namespace {
```

Comment:
what does this do ?

Reply:
Nothing currently?

uAMQP (and this code) currently does not care about the order of appearance for AMQP fields in a message. However the spec ambiguously specifies that there is a specific order to the fields. This vector is the expected order of the fields in a message - I was toying with the idea of validating that they appeared in the correct order but didn't actually write that code.